accepted/2023/feature-switch-attributes.md

```diff
@@ -0,0 +1,766 @@
+
+- Teach the ILLink Roslyn analyzer to treat `IsDynamicCodeCompiled` as a guard for `RequiresDynamicCodeAttribute` using the same model we define for third-party libraries
+
+### Non-goals
+
+- Support branch elimination in the analyzer for all feature switches
+
+  The most important use case for the analyzer is analyzing libraries. Libraries typically don't bake in constants for feature switches, so the analyzer needs to consider all branches. It should support feature guards for features that produce warnings, but doesn't need to consider feature settings passed in from the project file to treat some branches as dead code.
+
+- Teach the ILLink Roslyn Analyzer about the substitution XML
+
+  We don't want to teach the analyzer to read the substitution XML. The analyzer is the first interaction that users typically have with trimming and AOT warnings. This should not be burdened by the XML format. Even if we did teach the analyzer about the XML, it would not solve the problem; for example, the analyzer must not globally assume that `IsDynamicCodeSupported` is false as ILCompiler does.
+
+- Define a model with the full richness of the supported OS platform attributes
+
+  We will focus initially on a model where feature switches are booleans that return `true` if a feature is enabled. We aren't considering supporting version checks, or feature switches of the opposite polarity (where `true` means a feature is disabled/unsupported). We will consider what this might look like just enough to gain confidence that our model could be extended to support these cases in the future, but won't design this fully in the first iteration.
+
+## Feature guard attribute
+
+In order to treat a property as a guard for a feature that has a `Requires` attribute, there must be a semantic tie between the guard property and the attribute. ILLink and ILCompiler don't have this requirement because they run on apps, not libraries, so the desired warning behavior just falls out, thanks to substitution XML, branch elimination and the fact that `IsDynamicCodeSupported` is set to false from MSBuild.
+
+We could allow placing `FeatureGuardAttribute` on the property to indicate that it should act as a guard for a particular feature. The intention with any of these approaches is for the guard to prevent analyzer warnings:
+
+```csharp
+if (Feature.IsSupported) {
+    APIWhichRequiresDynamicCode(); // No warnings
+}
+
+[RequiresDynamicCode("Does something with dynamic codegen")]
+static void APIWhichRequiresDynamicCode() {
+    // ...
+}
+```
+
+The attribute instance needs to reference the feature somehow, whether as:
+
+- a reference to the feature attribute:
+
+  ```csharp
+  class Feature {
+      [FeatureGuard(typeof(RequiresDynamicCodeAttribute))]
```

Comment:
Using a type seems unfortunate because that would require on attribute per feature and would disallow more complex shapes (eg an attribute with flags).